### DISCUSSION TITLE
感觉切换速度好慢;_;

### DISCUSSION THREAD

**Discussion by LoneFireBlossom** [Yang-Xijie/InputSourceSwitcher#19] [General]
用keyboard maestro或者kawa，利用系统的那个TISSelectInputSource切换的话很快，但是偶尔会有那个众所周知的bug。
用这个的话好像是能准确切换的，但是每次切换输入法都要等上大概一秒，感觉真的等不了= = Monterey 12.1，MBP16 intel
那个open的issue我没太看懂，好像是说这个app在Monterey上慢而且不可靠，但是在Big Sur上使用没问题？

**Comment by Yang-Xijie** [comment]
我简单说明一下吧。

在macOS 12 推出的时候，我第一时间对项目进行了更新，commit了几次，但是没想到macOS API发生了变化，所以app最核心的切换功能挂掉了。这个时候我给repo的简介加了no longer maintain.

就是说，现在仓库的后面几次commit是完全没有意义的，因为既没有办法在macOS 12 work起来，也没有办法在macOS 11 work起来。

如果你希望在macOS 11（Big Sur）上使用这个app，下载release中的 [v1.2](https://github.com/Yang-Xijie/InputSourceSwitcher/releases/tag/v1.2) 就可以了，对应的commit是d5f309e678f8e97d83da95e6facc5b4eada6b501

如果你使用的是最新的系统（macOS 12），抱歉，这个app怎么都work不起来。我已经停止维护了

**Comment by Yang-Xijie** [comment]
我现在用的是macOS 12，下了一下v1.3试了一下，ummm 所以现在是好了吗？我这边MacBook2019 Intel，macOS 12.1，切输入法功能正常，而且也没有你所说的延迟

之前12.0.1测试的时候会有bug，就是AppleScript里面的click明明执行了但是没作用

**Reply by Yang-Xijie** [reply]
不行，还是有bug。时好时坏

**Comment by Yang-Xijie** [comment]
我新在README里面添加了notice说了不会再维护了。macOS 12 的切换确实时好时坏，我目前没有解决的方法。

Applescript感觉不是一个很好的编程方式；我另一个与Music.app相关的Applescript有的时候也不work。

**Comment by LoneFireBlossom** [comment]
不知道你有没有看过这个，用karabiner-elements做的。这个只能在两个输入法之间切换。（似乎跟kawa的思路类似）
https://www.v2ex.com/t/565667
可是我需要输英、中、日三种语言（这是系统里语言的顺序）。
我想的是，以切换到日文输入法为例，我希望按一下右侧opt键稳定切换到日文输入法，
如果karabiner能实现捕获到按下右opt键时：
```
if alone
    if 现在是中文输入状态
        则把按键映射为按一下F17 // 在系统偏好设置里已经设置“select next source in input menu”快捷键为F17
    if 现在是英文输入状态
        则把按键映射为极其快速地按两下F17  
if held_down
    则把按键映射为右opt
```
这样的话就能保证稳定切换了。
（PS：我现在是用karabiner-elements把左&右 CMD、OPT映射为F键，然后用keyboard maestro设置的左侧CMD切换英文，右侧CMD切换中文，右侧OPT切换日文，所以切换CJKV有bug；左侧opt为select next source in input menu，用来在出bug的时候切换）

但是我没搜到karabiner能把「按一下按键」映射为「按两下按键」，搜“karabiner double tap”全都是把“按两下按键映射为按一下按键”的……所以只好作罢。（不知道应该怎么搜才能搜到相关讨论，而且我本来也挺菜的，搜到估计也不会搞）

**Reply by Yang-Xijie** [reply]
我的话还是等Apple给出这一块新API吧。Carbon deprecated，但是目前好像没有新的API来代替

**Comment by LoneFireBlossom** [comment]
这几天一直在用kawa来切换输入法，感觉比用keyboard maestro来设置要好。
用keyboard maestro切换的问题就是都知道的，按下快捷键切换中文，系统显示输入法是中文，结果实际上还是英文。基本上每天打字都会遇到。

用kawa的话没有这个问题，它的问题是我按下切换到中文/日语输入法的快捷键之后，不能瞬间开始打字，得稍微缓个0.1秒左右（总之就是别太快就行）。这个问题比用keyboard maestro来切换输入法的问题要轻太多了，所以我的输入体验提升了很多。
别的就没任何问题了。

另外想知道你现在是用什么来切换输入法的呢？

**Reply by Yang-Xijie** [reply]
系统自带的fn键，macOS12新功能，比之前方便很多了
我的话只用中日英三个，所以还好
但是还是比较怀念之前能用这个app的macOS11，切换起来真的超级顺畅

**Reply by LoneFireBlossom** [reply]
我用一个快捷键轮换切换输入法时，有一点让我挺难受：
我在一个app里打完字，切换到另一个app里准备打字总是直接就开始敲键盘。
比如，之前在A.app里输入日语，现在切换到B.app里要输入中文，敲了几个字才发现不对，还得把输入的字都删了，然后再换输入法，然后再输入。虽然用了input source pro，但体验还是不太舒服。比如，因为系统有英中日三种输入法，所以切换输入法有的时候要按一下快捷键，有的时候要按两下，体验有点割裂。
所以就转到每个输入法单独分配一个快捷键上了。

PS：我的fn键给了emoji。

**Reply by Yang-Xijie** [reply]
试了下kawa，简直救星！

**Reply by LoneFireBlossom** [reply]
用kawa的话得注意按旧版readme里来设置，见：https://github.com/zcbenz/kawa

> Thus, to activate the workaround above, the following restrictions should meet.
There is **at least one non-CJKV input source in the input source list**（比如英语就是non-CJKV的输入，也就是说输入法里得启用英语）
**The Select the previous input source shortcut should be enabled** and set to something
It can be set in System Preferences > Keyboard > Shortcuts > Input Sources

**Reply by Yang-Xijie** [reply]
我也没怎么设置直接下载就能用了，目前好像没什么问题！life saver
可以高效的学日语了！
感谢

**Reply by LoneFireBlossom** [reply]
不设置的话就会遇到和上面keyboard maestro设置那样的问题，显示当前输入法是中文，实际上还是英文。设置了就行了

**Reply by Yang-Xijie** [reply]
我待会儿看看 😃

**Reply by LoneFireBlossom** [reply]
我看你的这个项目的readme里有提到kawa，你之前没用过吗😂

**Reply by Yang-Xijie** [reply]
应该用过，但后来自己写好了程序之后就忘记了 😂

**Reply by GanZhiXiong** [reply]
macOS Monterey 12.1 使用 Kawa 1.1.0 从搜狗输入法切换到英文输入法没问题，但是从英文输入法切换到搜狗输入法却无效。你们有遇到过吗？

**Reply by Yang-Xijie** [reply]
这些切换输入法的软件都是在系统的输入法之间切换的，不支持第三方软件

**Reply by GanZhiXiong** [reply]
**刚在 Bug Sur 下 11.6.1 测试了，是支持第三方软件的。**
我在 12.1 下又测试了系统的拼音输入法，用 kawa 也是不能控制的。
切换输入法的时候 kawa 的通知有提示，但是输入法没有变。 
**所以感觉是系统问题。**
